File: src/engine/vertex_array.py
import numpy as np
import OpenGL.GL as GL


class VertexArray:

    def __init__(self):
        self.id = GL.glGenVertexArrays(1)
        GL.glBindVertexArray(self.id)

        self.buffers = []
        self.vertex_count = 0
        self.index_buffer = 0

    def add_attribute(self, attr_list: int, size: int, data: list):
        buffer = GL.glGenBuffers(1)
        GL.glBindBuffer(GL.GL_ARRAY_BUFFER, buffer)
        GL.glBufferData(GL.GL_ARRAY_BUFFER, np.array(data, dtype="float32"), GL.GL_STATIC_DRAW)

        GL.glVertexAttribPointer(attr_list, size, GL.GL_FLOAT, GL.GL_FALSE, 0, None)
        GL.glEnableVertexAttribArray(attr_list)

        self.buffers.append(buffer)

    def add_index_buffer(self, data: list):
        self.index_buffer = GL.glGenBuffers(1)
        GL.glBindBuffer(GL.GL_ELEMENT_ARRAY_BUFFER, self.index_buffer)
        GL.glBufferData(GL.GL_ELEMENT_ARRAY_BUFFER, np.array(data, dtype="uint32"), GL.GL_STATIC_DRAW)

        self.vertex_count = len(data)

        GL.glBindBuffer(GL.GL_ELEMENT_ARRAY_BUFFER, 0)

    def bind(self):
        GL.glBindVertexArray(self.id)

    def unbind(self):
        GL.glBindVertexArray(0)

    def dispose(self):
        GL.glDeleteVertexArrays(1, self.id)
        for buffer in self.buffers:
            GL.glDeleteBuffers(1, buffer)

    def enable_index_buffer(self):
        GL.glBindBuffer(GL.GL_ELEMENT_ARRAY_BUFFER, self.index_buffer)

File: src/graphics/renderer/opengl/vertex_array.cpp
#include <memory>
#include <cstddef>

#include <glad/glad.h>

#include "graphics/renderer/buffer_layout.h"
#include "graphics/renderer/opengl/vertex_array.h"
#include "graphics/renderer/opengl/buffer.h"
#include "other/logging.h"
#include "other/assert.h"

VertexArray::VertexArray(GLuint array)
    : array(array) {
    DEB_DEBUG("Created vertex array {}", array);
}

VertexArray::~VertexArray() {
    glDeleteVertexArrays(1, &array);

    DEB_DEBUG("Deleted vertex array {}", array);
}

std::shared_ptr<VertexArray> VertexArray::create() {
    GLuint array;
    glGenVertexArrays(1, &array);
    glBindVertexArray(array);

    return std::make_shared<VertexArray>(array);
}

void VertexArray::bind() {
    glBindVertexArray(array);
}

void VertexArray::unbind() {
    glBindVertexArray(0);
}

void VertexArray::add_buffer(std::shared_ptr<Buffer> buffer, const BufferLayout& layout) {
    ASSERT(layout.elements.size() > 0, "Invalid layout");

    glBindBuffer(GL_ARRAY_BUFFER, buffer->buffer);

    size_t offset = 0;

    for (unsigned int i = 0; i < layout.elements.size(); i++) {
        const VertexElement& element = layout.elements[i];

        switch (element.type) {
            case GL_FLOAT:
                glVertexAttribPointer(element.index, element.size, element.type, GL_FALSE,
                        layout.stride, (GLvoid*) offset);
                break;
            case GL_INT:
                glVertexAttribIPointer(element.index, element.size, element.type,
                        layout.stride, (GLvoid*) offset);
                break;
            default:
                REL_CRITICAL("Unknown element type, exiting...");
                exit(1);
        }

        glEnableVertexAttribArray(element.index);
        offset += element.size * VertexElement::get_size(element.type);
    }

    buffers.push_back(buffer);
}

void VertexArray::add_index_buffer(std::shared_ptr<IndexBuffer> index_buffer) {
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, index_buffer->buffer);
    this->index_buffer = index_buffer;
}
